File: packages/python/langchain_nuggets/middleware/action_context.py
"""Byte-exact ACP action-context binding (version 1)."""
from __future__ import annotations

import hashlib
import math
import re
from typing import Any, Dict, Optional

import rfc8785

from langchain_nuggets.middleware.types import AuthorityEvaluationRequest

ACTION_CONTEXT_VERSION_V1 = 1
ACTION_CONTEXT_DOMAIN_V1 = "nuggets.acp.action_context.v1"
PARAMETERS_DOMAIN_V1 = "nuggets.acp.parameters.v1"
INTENT_DOMAIN_V1 = "nuggets.acp.intent.v1"

_MAX_SAFE_INTEGER = 2**53 - 1
_SHA256_HEX = re.compile(r"^[0-9a-f]{64}$")
_DELEGATION_ID = re.compile(r"^[1-9][0-9]*$")


def _assert_json_value(value: Any, path: str, allow_floats: bool) -> None:
    if value is None or isinstance(value, (str, bool)):
        return
    if isinstance(value, int):
        if abs(value) > _MAX_SAFE_INTEGER:
            raise ValueError(f"unsafe integer at {path}")
        return
    if isinstance(value, float):
        if not math.isfinite(value):
            raise ValueError(f"non-finite number at {path}")
        if not allow_floats:
            raise ValueError(f"non-integer number at {path}")
        return
    if isinstance(value, list):
        for index, item in enumerate(value):
            _assert_json_value(item, f"{path}[{index}]", allow_floats)
        return
    if isinstance(value, dict):
        for key, item in value.items():
            if not isinstance(key, str):
                raise ValueError(f"non-string object key at {path}")
            _assert_json_value(item, f"{path}.{key}", allow_floats)
        return
    raise ValueError(f"unsupported value at {path}")


def _hash_jcs(domain: str, value: Any, *, allow_floats: bool) -> str:
    _assert_json_value(value, "$", allow_floats)
    canonical = rfc8785.dumps(value)
    return hashlib.sha256(domain.encode("utf-8") + b"\n" + canonical).hexdigest()


def hash_parameters_v1(parameters: Dict[str, Any]) -> str:
    return _hash_jcs(PARAMETERS_DOMAIN_V1, parameters, allow_floats=True)


def hash_intent_v1(intent: str) -> str:
    if not intent:
        raise ValueError("intent must not be empty")
    return _hash_jcs(INTENT_DOMAIN_V1, intent, allow_floats=False)


def assert_canonical_delegation_id(value: str) -> str:
    if not _DELEGATION_ID.fullmatch(value) or int(value) > _MAX_SAFE_INTEGER:
        raise ValueError("delegation_id must be a canonical positive safe integer string")
    return value


def _assert_sha256(value: str, field: str) -> None:
    if not _SHA256_HEX.fullmatch(value):
        raise ValueError(f"{field} must be lowercase SHA-256 hex")


def build_action_context_preimage_v1(
    request: AuthorityEvaluationRequest,
    *,
    environment: Optional[str] = None,
    agent_version: Optional[str] = None,
) -> Dict[str, Any]:
    action = request.action
    if not action.tool:
        raise ValueError("tool must not be empty")
    _assert_sha256(action.parameters_hash, "parameters_hash")
    preimage: Dict[str, Any] = {
        "action_context_version": ACTION_CONTEXT_VERSION_V1,
        "tool": action.tool,
        "parameters_hash": action.parameters_hash,
        "agent_id": request.agent_id,
        "controller_id": request.controller_id,
    }
    if action.target is not None:
        preimage["target"] = action.target
    if action.intent_hash is not None:
        _assert_sha256(action.intent_hash, "intent_hash")
        preimage["intent_hash"] = action.intent_hash
    if environment is not None:
        preimage["environment"] = environment
    if agent_version is not None:
        preimage["agent_version"] = agent_version
    if action.amount_minor is not None:
        preimage["amount_minor"] = action.amount_minor
    if action.currency is not None:
        preimage["currency"] = action.currency
    preimage["delegation_id"] = assert_canonical_delegation_id(request.delegation_id)
    return preimage


def compute_action_context_hash_v1(
    request: AuthorityEvaluationRequest,
    *,
    environment: Optional[str] = None,
    agent_version: Optional[str] = None,
) -> str:
    return _hash_jcs(
        ACTION_CONTEXT_DOMAIN_V1,
        build_action_context_preimage_v1(
            request, environment=environment, agent_version=agent_version
        ),
        allow_floats=False,
    )

File: packages/js/src/actionContext.ts
import { createHash } from "node:crypto";
import canonicalize from "canonicalize";
import type { ActionContext, AuthorityEvaluationRequest, ToolArgs } from "./types.js";

export const ACTION_CONTEXT_VERSION_V1 = 1 as const;
export const ACTION_CONTEXT_DOMAIN_V1 = "nuggets.acp.action_context.v1";
export const PARAMETERS_DOMAIN_V1 = "nuggets.acp.parameters.v1";
export const INTENT_DOMAIN_V1 = "nuggets.acp.intent.v1";

const SHA256_HEX = /^[0-9a-f]{64}$/;
const DELEGATION_ID = /^[1-9][0-9]*$/;

function assertJsonValue(value: unknown, path: string, allowFloats: boolean): void {
  if (value === null || typeof value === "string" || typeof value === "boolean") return;
  if (typeof value === "number") {
    if (!Number.isFinite(value)) throw new Error(`non-finite number at ${path}`);
    if (!allowFloats && !Number.isInteger(value)) throw new Error(`non-integer number at ${path}`);
    if (Number.isInteger(value) && !Number.isSafeInteger(value)) {
      throw new Error(`unsafe integer at ${path}`);
    }
    return;
  }
  if (Array.isArray(value)) {
    value.forEach((item, index) => assertJsonValue(item, `${path}[${index}]`, allowFloats));
    return;
  }
  if (typeof value === "object" && Object.getPrototypeOf(value) === Object.prototype) {
    for (const [key, item] of Object.entries(value as Record<string, unknown>)) {
      if (item === undefined) throw new Error(`undefined value at ${path}.${key}`);
      assertJsonValue(item, `${path}.${key}`, allowFloats);
    }
    return;
  }
  throw new Error(`unsupported value at ${path}`);
}

function hashJcs(domain: string, value: unknown, allowFloats: boolean): string {
  assertJsonValue(value, "$", allowFloats);
  const serialized = canonicalize(value);
  if (typeof serialized !== "string") throw new Error("canonicalization produced no output");
  return createHash("sha256").update(`${domain}\n${serialized}`, "utf8").digest("hex");
}

export function hashParametersV1(parameters: ToolArgs): string {
  return hashJcs(PARAMETERS_DOMAIN_V1, parameters, true);
}

export function hashIntentV1(intent: string): string {
  if (intent.length === 0) throw new Error("intent must not be empty");
  return hashJcs(INTENT_DOMAIN_V1, intent, false);
}

function assertSha256(value: string, field: string): void {
  if (!SHA256_HEX.test(value)) throw new Error(`${field} must be lowercase SHA-256 hex`);
}

export function assertCanonicalDelegationId(value: string): string {
  if (!DELEGATION_ID.test(value) || !Number.isSafeInteger(Number(value))) {
    throw new Error("delegation_id must be a canonical positive safe integer string");
  }
  return value;
}

export function buildActionContextPreimageV1(input: {
  action: ActionContext;
  agentId: string;
  controllerId: string;
  delegationId?: string;
  environment?: string;
  agentVersion?: string;
}): Record<string, unknown> {
  const { action } = input;
  if (!action.tool) throw new Error("tool must not be empty");
  assertSha256(action.parameters_hash, "parameters_hash");
  const preimage: Record<string, unknown> = {
    action_context_version: ACTION_CONTEXT_VERSION_V1,
    tool: action.tool,
    parameters_hash: action.parameters_hash,
    agent_id: input.agentId,
    controller_id: input.controllerId
  };
  if (action.target !== undefined) preimage.target = action.target;
  if (action.intent_hash !== undefined && action.intent_hash !== null) {
    assertSha256(action.intent_hash, "intent_hash");
    preimage.intent_hash = action.intent_hash;
  }
  if (input.environment !== undefined) preimage.environment = input.environment;
  if (input.agentVersion !== undefined) preimage.agent_version = input.agentVersion;
  if (action.amount_minor !== undefined) preimage.amount_minor = action.amount_minor;
  if (action.currency !== undefined) preimage.currency = action.currency;
  if (input.delegationId !== undefined) {
    preimage.delegation_id = assertCanonicalDelegationId(input.delegationId);
  }
  return preimage;
}

export function computeActionContextHashV1(
  request: AuthorityEvaluationRequest,
  extras?: { environment?: string; agentVersion?: string }
): string {
  return hashJcs(
    ACTION_CONTEXT_DOMAIN_V1,
    buildActionContextPreimageV1({
      action: request.action,
      agentId: request.agent_id,
      controllerId: request.controller_id,
      delegationId: request.delegation_id,
      environment: extras?.environment,
      agentVersion: extras?.agentVersion
    }),
    false
  );
}
